### PR TITLE
Issue #11001: False positive for Whitespace Before Paren After String with emojis

### LINKED ISSUE #11001: Version 9.x Reports Whitespace Before Paren After String with Emojis
I have read check documentation: https://checkstyle.org/config_whitespace.html#ParenPad
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ javac Test.java

/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
  "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
  "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="ParenPad"/>
  </module>
</module>

/var/tmp $ cat Test.java
public class Test {
    void emojiFamilyRunner(Consumer<String> testRunner, String text) {
        testRunner.accept("👩‍👩‍👧‍👧 " + text); // count as 12 units
    }

    void emojiFamilyWithSkinToneModifierRunner(Consumer<String> testRunner, String text) {
        testRunner.accept("👩🏻‍👩🏽‍👧🏾‍👦🏿 " + text); // count as 20 units
    }
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-9.2-all.jar -c config.xml Test.java
Starting audit...
[ERROR] C:\Users\Work\Documents\Playground\CheckstylePlayground\.\src\main\java\Test.java:8:44: ')' is preceded with whitespace. [ParenPad]
[ERROR] C:\Users\Work\Documents\Playground\CheckstylePlayground\.\src\main\java\Test.java:12:48: ')' is preceded with whitespace. [ParenPad]
Audit done.
```

For Windows users, please use `type` instead of `cat` and run

```cmd
set RUN_LOCALE="-Duser.language=en -Duser.country=US"
java %RUN_LOCALE% -jar checkstyle-9.2-all.jar -c config.xml Test.java
```

in place of the last 2 commands above.

---

I would have expected no errors to be reported in this case but it appears the emoji messes up with where the paren is being seen. This doesn't occur when using 8.45.1.

This appears to be similar to https://github.com/checkstyle/checkstyle/issues/10920.

### PR DESCRIPTION
**Pull request by nmancus1**
Closes #11001 .

Diff Regression config: https://gist.githubusercontent.com/nmancus1/8eb01d0872a0abb22f818b802750a470/raw/c1474267ec16c4e3e9d7b035f112ed2952737f79/config.xml

Diff report(Clean): https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/6032fab_2021155411/reports/diff/index.html

### CODE REVIEW

**Comment by nmancus1** [comment]
Github, generate report

**Comment by nmancus1** [comment]
It seems that `pitest-ant` is being flaky after https://github.com/checkstyle/checkstyle/commit/5fd5f721957689797312ca388902f27c20dbfa99 , I cannot reproduce this locally.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/AbstractParenPadCheck.java [hunk 1]
```diff
@@ -109,13 +110,14 @@ protected void processRight(DetailAST ast) {
...
+            final int[] codePoints = line.codePoints().toArray();
             if (option == PadOption.NOSPACE
-                && Character.isWhitespace(line.charAt(before))
+                && CommonUtil.isCodePointWhitespace(codePoints, before)
                 && !CommonUtil.hasWhitespaceBefore(before, line)) {
                 log(ast, MSG_WS_PRECEDED, CLOSE_PARENTHESIS);
             }
             else if (option == PadOption.SPACE
-                && !Character.isWhitespace(line.charAt(before))
+                && !CommonUtil.isCodePointWhitespace(codePoints, before)
```
Comment (strkkk) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/AbstractParenPadCheck.java:120]: minor thing - `CommonUtil.isCodePointWhitespace(codePoints, before)` can be extracted, no need to invoke it 2 times

**Comment by romani** [comment]
Protest ant is restarted.
If problem is persistent, we should update pitest shell script

**Comment by romani** [comment]
Pitest utils failed with coverage less than 100%. Restarted

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/AbstractParenPadCheck.java (hunk 1 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/AbstractParenPadCheck.java]: Done

**Comment by nmancus1** [comment]
> Pitest utils failed with coverage less than 100%. Restarted

This was legit failure, needed to cover `CommonUtil#isCodePointWhitespace`. I added some simple tests to corresponding UTs instead of adding more target tests.

**Comment by nmancus1** [comment]
Github, generate report

**Review by strkkk — APPROVED** [review]

**Review by romani — CHANGES_REQUESTED** [review]
Items:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    void emojiFamilyRunner1(Consumer<String> testRunner, String text) {
+        testRunner.accept("👩‍👩‍👧‍👧 " + text); // ok
+    }
+
+    void emojiFamilyWithSkinToneModifierRunner1(Consumer<String> testRunner, String text) {
+        testRunner.accept("👩🏻‍👩🏽‍👧🏾‍👦🏿 " + text); // ok
+    }
+
+    void emojiFamilyRunner2(Consumer<String> testRunner, String text) {
+        testRunner.accept("👩‍👩‍👧‍👧 " + text ); // violation
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java:25]: It took me a while to understand where is problem

Can you use please put violation message in trailing comment?
https://github.com/checkstyle/checkstyle/blob/b9630a6475bef3e95430bfa7ac3cf5cc28f3a9e5/src/main/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/messages.properties#L15

**Review by romani — CHANGES_REQUESTED** [review]
Items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/AbstractParenPadCheck.java (hunk 1 above)
Comment (nmancus1) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/AbstractParenPadCheck.java]: @strkkk please resolve item

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java (hunk 2 above)
Comment (nmancus1) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java:25]: This:

```java
public class InputParenPadCheckEmoji {
    void emojiFamilyRunner1(Consumer<String> testRunner, String text) {
        testRunner.accept("👩‍👩‍👧‍👧 " + text); // ok
    }

    void emojiFamilyWithSkinToneModifierRunner1(Consumer<String> testRunner, String text) {
        testRunner.accept("👩🏻‍👩🏽‍👧🏾‍👦🏿 " + text); // ok
    }

    void emojiFamilyRunner2(Consumer<String> testRunner, String text) {
        testRunner.accept("👩‍👩‍👧‍👧 " + text ); // violation, ')' is preceded with whitespace
    }

    void emojiFamilyWithSkinToneModifierRunner2(Consumer<String> testRunner, String text) {
        testRunner.accept("👩🏻‍👩🏽‍👧🏾‍👦🏿 " + text ); // violation, ')' is preceded with whitespace
    }

    void emojiFamilyRunner3(Consumer<String> testRunner, String text) {
        testRunner.accept( "👩‍👩‍👧‍👧 " + text); // violation, '(' is followed by whitespace
    }

    void emojiFamilyWithSkinToneModifierRunner3(Consumer<String> testRunner, String text) {
        testRunner.accept( "👩🏻‍👩🏽‍👧🏾‍👦🏿 " + text); // violation, '(' is followed by whitespace
    }

    void emojiFamilyWithSkinToneModifierRunnerTricky(Consumer<String> testRunner, String text) {
        testRunner.accept( "👩🏻‍👩🏽‍👧🏾‍👦🏿 "  + "ab cdefg" + "👩🏽‍👧🏾‍👦🏿 " + text ); // 2 violations
    } // '(' is followed by whitespace, ')' is preceded with whitespace ^
}
```

Causes issues with inline config parser:

```bash
Violation lines for /home/nick/IdeaProjects/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java differ.
unexpected (4): 25, 29, 33, 37
---
expected      : [41, 41]
but was       : [25, 29, 33, 37, 41, 41]

	at com.puppycrawl.tools.checkstyle.AbstractModuleTestSupport.verifyViolations(AbstractModuleTestSupport.java:427)
	at com.puppycrawl.tools.checkstyle.AbstractModuleTestSupport.verifyWithInlineConfigParser(AbstractModuleTestSupport.java:270)


```

No change made to check. It seems like if trailing comment has anything extra, parser fails.

Comments have been placed under lines with violation.

File: src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java [hunk 3]
```diff
@@ -489,6 +491,27 @@ public void testGetUriByFilenameFindsRelativeResourceOnClasspathPrefix() throws
...
+        final int[] codePoints = " 123".codePoints().toArray();
+        assertThat(CommonUtil.isCodePointWhitespace(codePoints, 1))
+                .isFalse();
+    }
+
+    @Test
+    public void testIsCodePointWhitespaceOutOfBounds() {
+        final int[] codePoints = " 123".codePoints().toArray();
+        assertThrows(ArrayIndexOutOfBoundsException.class,
+                () -> CommonUtil.isCodePointWhitespace(codePoints, 4));
```
[the remark this answers is not in the document]
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java]: This test was extraneous, I just removed it.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java (hunk 2 above)
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java:25]: Can you try escape single quote?

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java:25]: And remove `,`

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java:25]: Please look at other places where message is used

**Comment by romani** [comment]
GitHub, rebase

**Comment by romani** [comment]
Github, generate report

**Comment by romani** [comment]
Github, generate report

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java (hunk 2 above)
Comment (nmancus1) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java]: Mostly done, I could not make two different violations on the same line work, and could not find an example of this case.

**Comment by nmancus1** [comment]
Github, generate site

**Comment by nmancus1** [comment]
Github, generate site

**Comment by romani** [comment]
Github, generate site

**Comment by romani** [comment]
GitHub, rebase

**Comment by nmancus1** [comment]
From https://github.com/checkstyle/checkstyle/runs/4484785952?check_suite_focus=true:

```
Successfully rebased and updated refs/heads/issue-11001.
+ git push --force-with-lease fork issue-11001
remote: Permission to nmancus1/checkstyle.git denied to github-actions[bot].
```

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java (hunk 2 above)
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java]: Two violation do not have this, we do not know how to even make it for now.
Ok to keep without message

**Comment by romani** [comment]
GitHub, rebase

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java (hunk 2 above)
Comment (nmancus1) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java]: ok, done now. I think the comma with `'` afterwards was the problem.

**Comment by nmancus1** [comment]
Github, rebase

**Review by romani — CHANGES_REQUESTED** [review]
Last item:

File: src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java [hunk 4]
```diff
@@ -489,6 +490,20 @@ public void testGetUriByFilenameFindsRelativeResourceOnClasspathPrefix() throws
...
+        final int[] codePoints = " 123".codePoints().toArray();
+        assertThat(CommonUtil.isCodePointWhitespace(codePoints, 0))
+                .isTrue();
+    }
+
+    @Test
+    public void testIsCodePointWhitespaceFalse() {
+        final int[] codePoints = " 123".codePoints().toArray();
+        assertThat(CommonUtil.isCodePointWhitespace(codePoints, 1))
+                .isFalse();
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java:504]: This two tests are indentical on input data please combine them to one method.

**Review by romani — CHANGES_REQUESTED** [review]
Last item:

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java (hunk 2 above)
Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckEmoji.java]: Awesome, thanks a lot

File: src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java (hunk 4 above)
Comment (nmancus1) [reply src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java:499]: Done.

**Review by romani — APPROVED** [review]
Ok to merge

File: src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java [hunk 5]
```diff
@@ -489,6 +490,15 @@ public void testGetUriByFilenameFindsRelativeResourceOnClasspathPrefix() throws
         assertEquals("Checker", config.getName(), "Unexpected config name!");
     }
 
+    @Test
```
Comment (nmancus1) [inline src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java:493]: Here

**Pull request merged**